### PR TITLE
Update the stub code of BOSS

### PR DESCRIPTION
**Pull request by JamePeng**
There are some wrong FunctionTable data in 3dbrew's BOSS service page, so I disassembled the boss lib~ to find the correct address and its function name.
Otherwise, I only update a little stub code to check the value.

image

### CODE REVIEW

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-08-26T12:41:56Z: https://github.com/citra-emu/citra/compare/5b9e41bf1c891a36526a0d8844fb40bbc3b1e870...JamePeng:e465c0662d645a4e39c003d7eddf0024096b1237

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-08-26T13:01:39Z: https://github.com/citra-emu/citra/compare/5b9e41bf1c891a36526a0d8844fb40bbc3b1e870...JamePeng:9d2a4827bb867c077610ee970168ddcec8b044cb

File: src/core/hle/service/boss/boss.h [hunk 1]
```diff
@@ -4,9 +4,138 @@
...
 namespace Service {
 namespace BOSS {
 
+/**
+ * BOSS::InitializeSession service function
+ *  Inputs:
+ *      0 : Header Code[0x00010082]
+ *      1 : u32 lower 64bit value (unused)
+ *      2 : u32 higher 64bit value (unused)
+ *      3 : u32 unknown traslation (if not 0x20 , The errorcode 0xD9001830 will set into cmd_buf[1])
```
Comment (wwylele) [inline src/core/hle/service/boss/boss.h]: 0x20 is always [CallingPidDesc](https://github.com/citra-emu/citra/blob/master/src/core/hle/kernel/session.h#L32) and the next one is a placeholder for PID (which citra doesn't care).

File: src/core/hle/service/boss/boss.h [hunk 2]
```diff
@@ -4,9 +4,138 @@
...
+ */
+void UpdateTaskCount(Service::Interface* self);
+
+/**
+ * BOSS::GetTaskState service function
+ *  Inputs:
+ *      0 : Header Code[0x00200082]
+ *      1 : u32 unknown_size
+ *      2 : 0
+ *      3 : u32 (unknown_translation & 0xF) == 0xA
```
Comment (wwylele) [inline src/core/hle/service/boss/boss.h]: ditto, [MappedBufferDesc(permission = R)](https://github.com/citra-emu/citra/blob/master/src/core/hle/kernel/session.h#L49)

File: src/core/hle/service/boss/boss.cpp [hunk 3]
```diff
@@ -2,19 +2,169 @@
...
+
+    cmd_buff[0] = IPC::MakeHeader(0x8, 0x1, 0);
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+
+    LOG_WARNING(Service_BOSS, "(STUBBED) unk_param1=0x%08X, unk_param2=0x%08X",unk_param1, unk_param2);
+}
+
+void SetOptoutFlag(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    output_flag = cmd_buff[1];
```
Comment (mailwl) [inline src/core/hle/service/boss/boss.cpp]: it most likely BOOL value (or, at least u8)

File: src/core/hle/service/boss/boss.cpp [hunk 4]
```diff
@@ -2,19 +2,169 @@
...
+                unk_param1, unk_size, unk_translation, unk_addr);
+}
+
+void UpdateTaskCount(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    // TODO(JamePeng): Figure out the meaning of these parameters
+    u32 unk_param1 = cmd_buff[1];
+    u32 unk_param2 = cmd_buff[2];
+    u32 unk_param3 = cmd_buff[3];
+    u32 unk_param4 = cmd_buff[4];
```
Comment (mailwl) [inline src/core/hle/service/boss/boss.cpp]: some buffer addr

File: src/core/hle/service/boss/boss.cpp [hunk 5]
```diff
@@ -2,19 +2,169 @@
...
+    cmd_buff[3] = 0; // stub 0 (16 bit value)
+
+    LOG_WARNING(Service_BOSS, "(STUBBED) unk_param1=0x%08X, unk_size=0x%08X, unk_translation=0x%08X, unk_addr=0x%08X",
+                unk_param1, unk_size, unk_translation, unk_addr);
+}
+
+void UpdateTaskCount(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    // TODO(JamePeng): Figure out the meaning of these parameters
+    u32 unk_param1 = cmd_buff[1];
```
Comment (mailwl) [inline src/core/hle/service/boss/boss.cpp]: size of buffer

File: src/core/hle/service/boss/boss.cpp [hunk 6]
```diff
@@ -2,19 +2,169 @@
...
+    u32 unk_param2 = cmd_buff[2];
+    u32 unk_param3 = cmd_buff[3];
```
Comment (mailwl) [inline src/core/hle/service/boss/boss.cpp]: 16 * (size of buffer) | 0xA;

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-08-26T16:50:51Z: https://github.com/citra-emu/citra/compare/b879d8c31b12662e787f08c8d8ab05115db8cc45...JamePeng:bb9d99cf1d883949310330ac09cdfda4a644ef0c

File: src/core/hle/service/boss/boss.cpp (hunk 4 above)
Comment (JamePeng) [inline src/core/hle/service/boss/boss.cpp]: mailwl. are you sure the meaning of the cmd_buf[4] ?

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-08-26T17:11:09Z: https://github.com/citra-emu/citra/compare/b879d8c31b12662e787f08c8d8ab05115db8cc45...JamePeng:e87b57d3340f118517e514ffea2bfac3a89db0fc

File: src/core/hle/service/boss/boss.cpp (hunk 6 above)
Comment (JamePeng) [inline src/core/hle/service/boss/boss.cpp]: ?

File: src/core/hle/service/boss/boss.cpp (hunk 4 above)
Comment (mailwl) [inline src/core/hle/service/boss/boss.cpp]: yes, it is https://www.3dbrew.org/wiki/IPC
Buffer Mapping Translation, in part:

0x0000000A | (size<<4)
<ptr>

The corresponding value contains a ptr to a buffer of said size. It is mapped read-only in the destination process.

File: src/core/hle/service/boss/boss.cpp (hunk 6 above)
Comment (mailwl) [inline src/core/hle/service/boss/boss.cpp]: see above

File: src/core/hle/service/boss/boss.cpp (hunk 4 above)
Comment (JamePeng) [inline src/core/hle/service/boss/boss.cpp]: Strange? It's just that I thought you said the function GetOwnNsDataIdList
image

**Pull request closed without merging**